doc/src/manual/style-guide.md

```diff
@@ -413,3 +413,56 @@ julia> h(1)
 
 Thus, use `Int` literals when possible, with `Rational{Int}` for literal non-integer numbers,
 in order to make it easier to use your code.
+
+## In long-form function definitions, prefer explicit `return` over implicit return values
+
+Explicit `return` on an expression makes it clear that the expression is intended to be returned and
+is not only used for its side effect. It also makes it possible to locally see that a certain sub-expression
+in a larger expression is returned while for an implicit return value one needs to analyze the whole outer expression.
+In cases where the return value of the function is irrelevant, use a `return` statement with no value.
+This can help the compiler since it doesn't have to reason about the type of the return value.
+
+For example, prefer
+
+```jl
+function f(x)
+    if x > 2
+        return "a"
+    else
+        return "b"
+    end
+end
+
+function g(x)
+    y = compute(x)
+    return compute(y)
+end
+
+function h!(x)
+    fill!(x, compute(x))
+    modify!(x)
+    return
```

Comment:
Can we find an example that doesn't run counter to other suggestions in this file?

> It is typical for such functions to also return the modified array for convenience.

I'm not the biggest fan of the first example either as I try to only use one return point if possible.  That's not a documented guideline, though.

Reply:
Agreed on the first point.

Regarding one return point, I personally think early return is much better. There is seldom need to do a bunch of cleanup in Julia so the advantage of having only one return point seems much reduced compared to e.g. C. If your comment was just about the specific example, I just took the simplest I could figure out, of course, you would typically just write the whole thing there as `f(x) = x > 2 ? "a" : "b"`. Can maybe come up with something better.

Reply:
Oh I love early returns (guard clauses) at the top of a function, too, but I'll write those as short-circuits.  I don't love returns in the complicated portion of an algorithm if I can help it.  In a short example though it's not clear which is which. What you have is just fine.